modules/v2x/v2x_proxy/proto_adapter/proto_adapter.cc

```diff
@@ -0,0 +1,230 @@
+  for (int idx_road = 0; idx_road < obu_light.road_trafficlight_size();
+       idx_road++) {
+    const auto &obu_road_light1 = obu_light.road_trafficlight(idx_road);
+    // Set the road index for lane
+    apollo::v2x::RoadTrafficLight_Attribute tl_attr =
+        apollo::v2x::RoadTrafficLight_Attribute_EAST;
+    if (obu_road_light1.road_direction() == 1) {
+      tl_attr = apollo::v2x::RoadTrafficLight_Attribute_EAST;
+    } else if (obu_road_light1.road_direction() == 2) {
+      tl_attr = apollo::v2x::RoadTrafficLight_Attribute_WEST;
+    } else if (obu_road_light1.road_direction() == 3) {
+      tl_attr = apollo::v2x::RoadTrafficLight_Attribute_SOUTH;
+    } else if (obu_road_light1.road_direction() == 4) {
+      tl_attr = apollo::v2x::RoadTrafficLight_Attribute_NORTH;
+    } else {
+      AINFO << "Road direction=" << obu_road_light1.road_direction()
+            << " is invalid.";
+    }
+    // FOR-EACH LANE
+    for (int idx_lane = 0; idx_lane < obu_road_light1.lane_trafficlight_size();
+         idx_lane++) {
+      const auto &obu_lane_light1 = obu_road_light1.lane_trafficlight(idx_lane);
+      if (!obu_lane_light1.has_gps_x_m() || !obu_lane_light1.has_gps_y_m()) {
+        AWARN << "Invalid lane_trafficlight: [" << idx_road << "][" << idx_lane
+              << "]: no gps info here.";
+        continue;
+      }
+      flag_has_data = true;
+      auto res_light1_ptr = res->add_road_traffic_light();
+      res_light1_ptr->set_gps_x_m(obu_lane_light1.gps_x_m());
+      res_light1_ptr->set_gps_y_m(obu_lane_light1.gps_y_m());
+      res_light1_ptr->set_road_attribute(tl_attr);
+      // single traffic light
+      for (int j = 0; j < obu_lane_light1.single_traffic_light_size(); j++) {
+        auto res_single1_ptr = res_light1_ptr->add_single_traffic_light();
+        const auto &obu_single1 = obu_lane_light1.single_traffic_light(j);
+        if (obu_single1.has_id()) {
+          res_single1_ptr->set_id(obu_single1.id());
+        }
+        if (obu_single1.has_color_remaining_time_s()) {
+          res_single1_ptr->set_color_remaining_time_s(
+              obu_single1.color_remaining_time_s());
+        }
+        if (obu_single1.has_next_remaining_time()) {
+          res_single1_ptr->set_next_remaining_time_s(
+              obu_single1.next_remaining_time());
+        }
+        if (obu_single1.has_right_turn_light()) {
+          res_single1_ptr->set_right_turn_light(obu_single1.right_turn_light());
+        }
+        if (obu_single1.has_color()) {
+          res_single1_ptr->set_color(obu_single1.color());
+        }
+        if (obu_single1.has_next_color()) {
+          res_single1_ptr->set_next_color(obu_single1.next_color());
+        }
+        if (obu_single1.has_traffic_light_type()) {
+          res_single1_ptr->add_traffic_light_type(
+              AdapterLightTypeObu2Sys(obu_single1.traffic_light_type()));
+        }
+      }
+    }
+  }
+  *os_light_ptr = res;
+  return flag_has_data;
```

Comment:
I personally think `has_data` is enough to indicate it's a flag.

Reply:
Here is _repeated-in-repeated_ , so a single ```has_data``` cannot be used to complete the judgment. The logic here is that as long as the result contains any valid data, it should return true.